changelog/ImportC.md

```diff
@@ -0,0 +1,237 @@
+be translated into D.
+Although hand translating the .h files to D is not difficult,
+it is tedious, annoying, and definitely a barrier to using D
+with existing C code.
+
+Why can't the
+D compiler simply read the .h file and extract its declarations?
+Why doesn't it "just work"?
+D has had great success with integrating documentation
+generation into the language, as well as unit testing. Despite the
+existence of many documentation generators and testing frameworks,
+the simplicity of it being built in and "just working" is transformative.
+
+## The C Preprocessor
+
+Is its own language that is completely distinct from the C language
+itself. It has its own grammar, its own tokens, and its own rules.
+The C compiler is not even aware of the existence of the preprocessor.
+(The two can be integrated, but that doesn't change the fact
+that the two semantically know nothing about each other.)
+The preprocessor is different enough from D that there's no hope
+of translating preprocessor directives to D in anything but the most
+superficial manner, any more than the preprocessor directives can
+be replaced with C.
+
+## Previous Solutions
+
+### htod by Walter Bright
+
+[htod](https://dlang.org/htod.html) converts a C .h file
+to a D source file, suitable for importing into D code.
+htod is built from the front end of the Digital Mars C and C++ compiler.
+It works just like a C or C++ compiler except that its output is source
+code for a D module rather than object code.
+
+### DStep by Jacob Carlborg
+
+[DStep code](https://code.dlang.org/packages/dstep)
+
+[DStep Article](https://dlang.org/blog/2019/04/22/dstep-1-0-0/)
+
+From the Article: "DStep is a tool for automatically generating D
+bindings for C and Objective-C libraries. This is implemented by
+processing C or Objective-C header files and outputting D modules.
+DStep uses the Clang compiler as a library (libclang) to process the header files."
+
+### dpp by Átila Neves
+
+[dpp code](https://code.dlang.org/packages/dpp/0.2.1)
+
+[dpp Article](https://dlang.org/blog/2019/04/08/project-highlight-dpp/)
+
+From the Article: "dpp is a compiler wrapper that will parse a D source
+file with the .dpp extension and expand in place any #include directives
+it encounters, translating all of the C or C++ symbols to D, and then
+pass the result to a D compiler (DMD by default)."
+
+Like DStep, dpp relies on libclang.
+
+## Introducing ImportC, an ANSI C11 Compiler
+
+Here is the next step:
+
+* Forget about the C preprocessor.
+* Overlook C++.
+* Put a real ANSI C11 compiler in the D front end.
+* Call it ImportC to distinguish this unique capability.
+
+In detail:
+
+1. Compile C code directly, but *only* C code that has already been run through
+the C preprocessor. To import stdio.h into a D program,
+the build script would be:
+
+     gcc -E stdio.h >stdio.c
+
+and in the D source file:
+
+     import stdio;  // reads stdio.c and compiles it
```

Comment:
And how does it determine the C module name? Does it ever go into a package namespace? How does this interact with the import path?

Reply:
2. Do we have a `pragma` instruction to fix this ? something like `pragma(importFile, "wayland-util.h")` would be appreciated.

Reply:
> 2. Do we have a `pragma` instruction to fix this ? something like `pragma(importFile, "wayland-util.h")` would be appreciated.

I'm not thinking too hard about it, but there are features in the language that don't necessarily mean we need yet another feature added.

```
module c.wayland_util;
mixin(import("wayland-util.h"));
```

Reply:
> > 1. Do we have a `pragma` instruction to fix this ? something like `pragma(importFile, "wayland-util.h")` would be appreciated.
> 
> I'm not thinking too hard about it, but there are features in the language that don't necessarily mean we need yet another feature added.
> 
> ```
> module c.wayland_util;
> mixin(import("wayland-util.h"));
> ```

Oh nice. However, `import("filename");` doesn't seem to be documented [here](https://dlang.org/spec/module.html#import-declaration).

Reply:
It's documented [here](https://dlang.org/spec/expression.html#import_expressions).

Reply:
> 1. .d is looked for, then .c

Note that this only applies to one import path, not the entire list. The following code will be changed by this patch:

```
Project directory:

main/
  - app.d
  - stuff.c
util/
  - stuff.d
```

```d
// main/app.d

import stuff;

void main()
{
    import std.stdio : writeln;
    writeln(foo());
}
```

```c
// main/stuff.c
int foo()
{
    return 1;
}
```

```d
// util/stuff.d
int foo()
{
    return 2;
}
```

DMD:
```
dmd -Imain -Iutil -i main/app.d
```

The resulting binary will now print `1` instead of `2`.

---

THis project layout might be uncommon but at least warrants a warning in the changelog

Reply:
Case in point:
- Bindings: https://github.com/bosagora/agora/tree/4154e6eb477b5eab1e6effaa57fb5dda8ad5bc1a/source/scpd/scp
- Source: https://github.com/bosagora/agora/tree/4154e6eb477b5eab1e6effaa57fb5dda8ad5bc1a/source/scpp/src/scp

Same layout, just different folder, everything under `source`. In our case it's C++ code so unaffected, but @MoonlightSentinel 's point is sensible.

Reply:
> Case in point

There are no .d files in src/scp, so why would existing code put it in the import path for D files?

Reply:
> The resulting binary will now print 1 instead of 2.

Even if someone did manage to put a .c file with the same name as a .d file in a later directory along the D import path, the odds are about 99.99% it will fail to compile as ImportC won't deal with #include. Even if it did, to not get a compile error the .c file function prototype would have to match the one in the .d one.

Reply:
Sure, it's unlikely that such code would compile and exhibit different behaviour. But the real issue here is that this PR introduces a breaking change for existing projects. 
(The silent change in behaviour was just a worst case scenario)